File: frontend/client/app/services/categoryService.js
import BaseService from './baseService';

export default class CategoryService extends BaseService {
	all() {
		return super.get('categories/all');
	}

	create(category) {
		return super.post('categories', category);
	}

    getPage(page) {
        return super.get('categories/page/'+page);
    }

    delete(id) {
		return super.delete('categories/' + id);
	}

    find(id) {
        return super.get('categories/' + id);
    }

    update(id, category) {
        return super.post('categories/' + id, category);
    }

    filterByName(name) {
        return super.get('categories/filter-by/name?name=' + name);
    }
}

File: src/main/java/com/timxyz/services/CategoryService.java
package com.timxyz.services;

import com.timxyz.models.Category;
import com.timxyz.repositories.CategoryRepository;
import com.timxyz.services.exceptions.ServiceException;
import org.springframework.stereotype.Service;

import java.util.Collection;

@Service
public class CategoryService extends BaseService<Category, CategoryRepository> {

    public Category save(Category model) throws ServiceException {
        Category sameName = getByName(model.getName());

        if (sameName != null && model.getId() != sameName.getId()) {
            throw new ServiceException("Kategorija s navedenim imenom već postoji.");
        }

        if (model.getId() != null) {
            // Check against cyclic parent categories
            Category parent = model.getParent();

            while (parent != null) {
                if (parent.getId() == model.getId()) {
                    throw new ServiceException("Uočena je ciklična veza između nadkategorija.");
                }

                parent = parent.getParent();
            }
        }

        return super.save(model);
    }

    public void delete(Long id) throws ServiceException {
        Category category = get(id);

        if (category.getItems().size() > 0) {
            throw new ServiceException("Kategorija se ne može obrisati jer postoje inventurne stavke koje joj pripadaju.");
        }

        super.delete(id);
    }

    public Collection<Category> filterByName(String name) {
        return repository.filterByName(name);
    }

    public Category getByName(String name) {
        return repository.findFirstByName(name);
    }
}
